### PR TITLE
Create foundation button components

### PR DESCRIPTION
**Pull request by tkgnm**
#### Description

**Figma**: https://guardian.github.io/storybooks/?path=/story/source_react-components-button--primary-priority-default-theme

This PR adds our design systems foundation "button" component to the Swift source package. For now I've only added one very standard "brand" theme for each of our button styles.

Crucially, leaving out every variant should not matter and was sort of deliberate. I want to make it easy for consumers of this button to apply any sort of styling they so wish via an extension to the public `ButtonTheme` struct. This mechanism will allow users of this component to create new styles without having to update this package, giving developers the flexibility to expand this package when they are ready. 

In the future, we may wish to "merge" the IconButton with the SourceButtonStyle so that we have one component that can be a button with just an icon, just text or both as per the design system. 

#### Checklist
- [x] Changes have been checked by the developer
- [ ] Changes have been checked by the reviewers
- [ ] Unit tested

##### Recommended reviewiers

- Design review for UI changes from @guardian/design-system
- Code review from @guardian/android-developers or @guardian/ios-developers
- Optional code/API review from @guardian/client-side-infra

#### For pull requests introducing UI changes:

- [ ] Sign-off by Design: 
- [x] Dark Mode 
- [x] Tablet 
- [x] Accessibility (e.g. VoiceOver): 

##### Screenshots or videos:

| Light | Dark | Accessibility | 
| --- | --- | --- |
|Screenshot 2024-10-30 at 11 02 37|Screenshot 2024-10-30 at 11 02 48|Screenshot 2024-10-30 at 11 03 10|

### CODE REVIEW

**Comment by ab-gnm** [comment]
If you want to add all sizes and colour variants, they are described here: https://www.figma.com/design/b2qv2OMLoNCYnP01ipfrP7/%E2%97%88-Core-library?node-id=4225-149&m=dev

The dark mode colours are defined here: https://www.figma.com/design/2aPypHwKuPZkMgYuWrC7Qo/Dark-mode---DS-components?node-id=69-14454&m=dev

This Android PR has screenshots of all variants on Android: https://github.com/guardian/source-apps/pull/44

One final point: when naming button styles, using `priority` as suggested by [Simon here](https://github.com/guardian/source-apps/pull/44#discussion_r1617318234) has been very useful. The components in Figma are defined with those two values, e.g. "Primary" and "onWhite". This allows developers to immediately use them in the code.

As an example, here are the Figma and our code for a button:

| Header | Header |
|--------|--------|
| Screenshot 2024-10-24 at 10 58 45 | Screenshot 2024-10-24 at 11 00 17 |

**Comment by tkgnm** [comment]
> If you want to add all sizes and colour variants, they are described here: https://www.figma.com/design/b2qv2OMLoNCYnP01ipfrP7/%E2%97%88-Core-library?node-id=4225-149&m=dev
> 
> The dark mode colours are defined here: https://www.figma.com/design/2aPypHwKuPZkMgYuWrC7Qo/Dark-mode---DS-components?node-id=69-14454&m=dev
> 
> This Android PR has screenshots of all variants on Android: #44
> 
> One final point: when naming button styles, using `priority` as suggested by [Simon here](https://github.com/guardian/source-apps/pull/44#discussion_r1617318234) has been very useful. The components in Figma are defined with those two values, e.g. "Primary" and "onWhite". This allows developers to immediately use them in the code.
> 
> As an example, here are the Figma and our code for a button:
> 
> Header	Header
> Screenshot 2024-10-24 at 10 58 45	Screenshot 2024-10-24 at 11 00 17

Thanks Adi - the dark mode colours are very useful! 

With regards to naming, this PR indeed makes use of button priority. I've also used "ButtonTheme" which I kind of stole directly from the [csnx repo](https://github.com/guardian/csnx/blob/986f38f426c9b577200455e0b67e43c3577ef2e8/libs/%40guardian/source/src/react-components/button/theme.ts#L4). Let me know if you can think of any refinements?

File: Sources/Source/Components/Buttons/ButtonStyle+SourceButtonStyle.swift [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+            }
+        }
+    }
+}
+
+extension ButtonStyle where Self == SourceButtonStyle {
+    static func source(size: ButtonSize, priority: ButtonPriority, theme: ButtonTheme) -> SourceButtonStyle {
+        SourceButtonStyle(buttonSize: size, buttonPriority: priority, buttonTheme: theme)
+    }
```
Comment (aoifemcl15) [inline Sources/Source/Components/Buttons/ButtonStyle+SourceButtonStyle.swift:64]: This will need to be public so that is accessible in other projects.

**Review by aoifemcl15 — CHANGES_REQUESTED** [review]
Looks great - just a comment regarding access control of the styling modifier.

File: Sources/Source/Components/Buttons/ButtonStyle+SourceButtonStyle.swift (hunk 1 above)
Comment (tkgnm) [reply Sources/Source/Components/Buttons/ButtonStyle+SourceButtonStyle.swift:64]: Good spot! Thanks. Have made that adjustment. Had to make a couple other things public too as a result. Still waiting for sign off from design, so will hold off merging until I've got the all clear from them.

**Review by aoifemcl15 — APPROVED** [review]

**Pull request merged**